libsolidity/lsp/Transport.cpp

```diff
@@ -0,0 +1,136 @@
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#include <libsolidity/lsp/Transport.h>
+
+#include <libsolutil/JSON.h>
+#include <libsolutil/Visitor.h>
+#include <libsolutil/CommonIO.h>
+#include <liblangutil/Exceptions.h>
+
+#include <boost/algorithm/string.hpp>
+
+#include <iostream>
+#include <sstream>
+
+using namespace std;
+using namespace solidity::lsp;
+
+JSONTransport::JSONTransport(istream& _in, ostream& _out):
+	m_input{_in},
+	m_output{_out}
+{
+}
+
+JSONTransport::JSONTransport():
+	JSONTransport(cin, cout)
+{
+}
+
+bool JSONTransport::closed() const noexcept
+{
+	return m_input.eof();
+}
+
+optional<Json::Value> JSONTransport::receive()
+{
+	auto const headers = parseHeaders();
+	if (!headers)
+	{
+		error({}, ErrorCode::ParseError, "Could not parse RPC headers.");
+		return nullopt;
+	}
+
+	if (!headers->count("content-length"))
+	{
+		error({}, ErrorCode::ParseError, "No content-length header found.");
+		return nullopt;
+	}
+
+	string const data = util::readBytes(m_input, stoul(headers->at("content-length")));
+
+	Json::Value jsonMessage;
+	string jsonParsingErrors;
+	solidity::util::jsonParseStrict(data, jsonMessage, &jsonParsingErrors);
+	if (!jsonParsingErrors.empty() || !jsonMessage || !jsonMessage.isObject())
+	{
+		error({}, ErrorCode::ParseError, "Could not parse RPC JSON payload. " + jsonParsingErrors);
+		return nullopt;
+	}
+
+	return {move(jsonMessage)};
+}
+
+void JSONTransport::notify(string _method, Json::Value _message)
+{
+	Json::Value json;
+	json["method"] = move(_method);
+	json["params"] = move(_message);
+	send(move(json));
+}
+
+void JSONTransport::reply(MessageID _id, Json::Value _message)
+{
+	Json::Value json;
+	json["result"] = move(_message);
+	send(move(json), _id);
+}
+
+void JSONTransport::error(MessageID _id, ErrorCode _code, string _message)
+{
+	Json::Value json;
+	json["error"]["code"] = static_cast<int>(_code);
+	json["error"]["message"] = move(_message);
+	send(move(json), _id);
+}
+
+void JSONTransport::send(Json::Value _json, MessageID _id)
+{
+	solAssert(_json.isObject());
+	_json["jsonrpc"] = "2.0";
+	_json["id"] = _id;
+
+	string const jsonString = solidity::util::jsonCompactPrint(_json);
+
+	m_output << "Content-Length: " << jsonString.size() << "\r\n";
+	m_output << "\r\n";
+	m_output << jsonString;
+
+	m_output.flush();
+}
+
+optional<map<string, string>> JSONTransport::parseHeaders()
+{
+	map<string, string> headers;
+
+	while (true)
+	{
+		string line;
+		getline(m_input, line);
+		if (boost::trim_copy(line).empty())
+			break;
+
+		auto const delimiterPos = line.find(':');
+		if (delimiterPos == string::npos)
+			return nullopt;
+
+		string name = boost::to_lower_copy(line.substr(0, delimiterPos));
+		string value = line.substr(delimiterPos + 1);
+		headers[boost::trim_copy(name)] = boost::trim_copy(value);
```

Comment:
It's actually much more complicated:
```
 Multiple message-header fields with the same field-name MAY be present in a message if and only if the entire field-value for that header field is defined as a comma-separated list [i.e., #(values)]. It MUST be possible to combine the multiple header fields into one "field-name: field-value" pair, without changing the semantics of the message, by appending each subsequent field-value to the first, each separated by a comma. The order in which header fields with the same field-name are received is therefore significant to the interpretation of the combined field value, and thus a proxy MUST NOT change the order of these field values when a message is forwarded.
```

In addition to that, field values can be spread across multiple lines.

I would just ignore all of that for now and fix it if it causes a problem.

Reply:
BTW, for some context, this was my original question:

> What if the same header appears more than once? Currently it gets overwritten. Shouldn't it be an error instead?

I thought you were actually responding to something else before I found that comment :)